core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -866,7 +891,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * readFile("data.json").tap(putStrLn)
    * }}}
    */
-  final def tap[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, _]): ZIO[R1, E1, A] = self.flatMap(a => f(a).const(a))
+  final def tap[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, _]): ZIO[R1, E1, A] = self.flatMap(new ZIO.TapFn(f))
```

Comment:
All of this manual wrapping will be hard to maintain and easy to get wrong (plus, of course, it adds overhead that cannot be escaped).